v3/lints/cabf_br/lint_sub_ca_eku_incompatible_values.go

```diff
@@ -0,0 +1,94 @@
+/*
+ * ZLint Copyright 2021 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/*
+ * This lint contributed by Adriano Santoni <adriano.santoni@staff.aruba.it>
+ * of ACTALIS S.p.A. (an ARUBA company). Last revised June 30, 2021.
+ *
+ * Checks that the EKU extension in a Subordinate CA certificate meets the
+ * "MUST NOT" requirement set out in the BRs 7.1.2.2, letter g):
+ *
+ *     For Subordinate CA Certificates that will be used to issue
+ *     TLS certificates, the value id-kp-serverAuth[RFC5280] MUST be present.
+ *     The value id-kp-clientAuth[RFC5280] MAY be present.
+ *     The values id-kp-emailProtection[RFC5280], id-kp-codeSigning[RFC5280],
+ *     id-kp-timeStamping[RFC5280], and anyExtendedKeyUsage[RFC5280] MUST NOT
+ *     be present. Other values SHOULD NOT be present.
+ *     For Subordinate CA Certificates that are not used to issue TLS certificates,
+ *     then the value id-kp-serverAuth[RFC5280] MUST NOT be present.
+ *     Other values MAY be present, but SHOULD NOT combine multiple independent usages
+ *     (e.g.including id-kp-timeStamping[RFC5280] with id-kp-codeSigning[RFC5280]).
+ */
+
+package cabf_br
+
+import (
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_sub_ca_eku_incompatible_values",
+		Description:   "Subordinate CA extkeyUsage: if serverAuth is present, then emailProtection, codeSigning, timeStamping, and anyExtendedKeyUsage MUST NOT be present.",
+		Citation:      "BRs: 7.1.2.2",
+		Source:        lint.CABFBaselineRequirements,
+		EffectiveDate: util.CABFBRs_1_7_1_Date,
+		Lint:          NewSubCAEkuIncompatibleValues,
+	})
+}
+
+type subCAEkuIncompatibleValues struct{}
+
+func NewSubCAEkuIncompatibleValues() lint.LintInterface {
+	return &subCAEkuIncompatibleValues{}
+}
+
+func (l *subCAEkuIncompatibleValues) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubCA(c) && util.IsExtInCert(c, util.EkuSynOid)
+}
+
+func (l *subCAEkuIncompatibleValues) Execute(c *x509.Certificate) *lint.LintResult {
+
+	serverAuthPresent := false
+	emailProtectionPresent := false
+	codeSigningPresent := false
+	timeStampingPresent := false
+	anyEkuPresent := false
+
```

Comment:
Note that 7.1.2.2(g) also includes this requirement:

> For Subordinate CA Certificates that are not used to issue TLS certificates, then the value id-kp-serverAuth [RFC5280] MUST NOT be present. Other values MAY be present, but SHOULD NOT combine multiple independent key purposes (e.g. including id-kp-timeStamping [RFC5280] with id-kp-codeSigning [RFC5280]).

This code doesn't seem to test for that. I think it should, right?

Reply:
That's right, Ryan. My lint is specifically targeting the following part of BR 7.1.2.2(g):

`For all other Subordinate CA Certificates, including Technically ConstrainedSubordinate CA Certificates:`
`For Subordinate CA Certificates that will be used to issue TLS certificates, the valueid-kp-serverAuth[RFC5280] MUST be present. The valueid-kp-clientAuth[RFC5280] MAY be present. The valuesid-kp-emailProtection[RFC5280],id-kp-codeSigning[RFC5280],id-kp-timeStamping[RFC5280], andanyExtendedKeyUsage[RFC5280]MUST NOT be present. `

It was not my intention to cover the entirety of BR 7.1.2.2(g), but only the case where the serverAuth kp is present.